src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+                        Files.write(Paths.get(speciesMicroarrayFile.toString()), probeIdLine.getBytes(), StandardOpenOption.APPEND);
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Attempts to create a connection to Biomart with the supplied query
+     * @param queryString String of Biomart query
+     * @return Set<String> Biomart response data.
+     * @throws Exception Can be caused by the query URL not being valid, or the data not existing for the species.
+     */
+    private Set<String> queryBiomart(String queryString, int retryCount) throws Exception {
+        // Create connection to Biomart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == 200) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                // Biomart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    retryCount++;
+                    if (retryCount < 5) {
+                        Thread.sleep(10000);
+                        logger.warn("Biomart query failed. Trying again...");
+                        return queryBiomart(queryString, retryCount);
+                    }
+                    throw new Exception("Biomart query failed with message: " + line);
+                } else {
+                    biomartResponseLines.add(line);
+                }
+            }
+            return biomartResponseLines;
+
+        } else {
+            retryCount++;
+            if (retryCount < 5) {
+                Thread.sleep(10000);
+                logger.warn("Unable to connect to Biomart. Trying again...");
+                return queryBiomart(queryString, retryCount);
+            }
+            throw new Exception("Unable to connect to Biomart with URL: " + queryString);
+        }
+    }
+
+    /**
+     * Function that takes species name attribute from config file (eg: Homo sapiens) and modifies it to
+     * match Biomart formatting (first letter from primary species name + secondary name, all lowercase -- eg: hsapiens).
+     * @return List<String> of species names in Biomart format (eg: hsapiens).
+     * @throws IOException
+     * @throws ParseException
+     */
+    private List<String> getBiomartSpeciesNames() throws IOException, ParseException {
```

Comment:
I could be wrong but I think you said it's the same species list as Orthopairs? Maybe this code could be extracted to release-common-lib if it's common to both Release steps.

Reply:
Also the same one used for Orthoinference, yea I think in the future extracting these things to a common method should happen

Reply:
Maybe create a separate issue to do that as future work, if it's too much for this PR.

Reply:
Created an issue in release-common-lib, but in the meantime I created a Utils class in AddLinks that uses this method